### ISSUE TITLE
addEventListener defer option to call that handler after all other handlers run

### ISSUE THREAD

**Issue by andyearnshaw** [whatwg/dom#730]
I'm adding a key handler to a behaviour that a consumer adds to an element to virtualise scrolling.  The handler intercepts the keys that normally scroll the element and adjusts the scroll value in certain situations.  I can't guarantee that the consumer will not add their handler before the behaviour is attached, which means that, if the consumer adds a handler and wants to cancel the behaviour action then I can't easily know that.  You also don't know whether the consumer might use event delegation to handle events further up the tree, so you'd have to use some unpalatable options (event delegation at the root, overriding `addEventListener()` to remove and re-add yours whenever another handler is added for that type) to do your default action.

I'm proposing that an option, `defer`, that tells the browser to execute that handler after the bubbling phase has run.  This lets custom elements define their own default actions without having to use custom events:

```javascript
element.addEventListener('keydown', event => {
    if (event.defaultPrevented) {
        return;
    }
 
    event.preventDefault();
    // do the custom behaviour
}, { defer: true });
```

The handler always executes even if `event.stopPropagation()` or `event.stopImmediatePropagation()` was called. 

*Note: I'm thinking this is mostly useful for custom elements, in fact my behaviour would be implemented as a custom element if it weren't for it being too small to require the consumers use a web components polyfill. I could propose a more custom element-y solution over at the web components repo, but I feel this may be useful for devs who want to make lightweight behaviours for elements, hence suggesting here first.*

**Comment by annevk** [comment]
What if the consumer adds the same handler? It seems either way you need to negotiate some kind of protocol.

**Comment by andyearnshaw** [comment]
Do you mean if they add the same callback function *or* if they bind to the "deferred phase" for the same event?  I presume it's the latter, but just in case, I would expect the former to just ignore the call since the callback already exists within the set of callbacks attached to that event for the deferred phase.  In the case of the consumer binding their own functions to an event in the deferred phase, for which a behaviour has already attached an event, I guess it would call the callbacks in the order that they were added, as is the case for the capturing and bubbling phases.

I did almost call out this potential issue in my original post, but I'm not sure if it's really an issue, and that we couldn't just make it clear in documentation that the deferred phase isn't intended for the majority of consumer event handlers.  Maybe that wouldn't stop it from being misused, though.

**Comment by annevk** [comment]
Sorry, I meant if they also add a deferred handler. And I mostly meant to point out that you always need to cooperate with each other to some extent and figure out how that works as there are no real guarantees the platform can give at this level.

**Comment by andyearnshaw** [comment]
Yeah, I understand that, I probably should have made it clearer from the start that the proposal is not looking for such guarantees.  This is really just a proposal for making it easier for the component to do its thing (probably) after everything else, and in particular make that cooperation more natural by allowing the consumer to use the familiar `preventDefault()` to tell a component not to do whatever it normally does.

**Comment by andyearnshaw** [comment]
After running into the problem I described, I came back here but realised that my original proposal wasn't particularly eloquent. So, I've tried to make this more of a discussion where a more concrete proposal might form naturally.
